### ISSUE TITLE
Make "New Project" Wizard inital selection a flat list

### ISSUE THREAD

**Issue by laeubi** [eclipse-platform/eclipse.platform.ui#503]
Currently if one choose New > Project ... Eclipse open a dialog with a tree of all "project" wizards.

![grafik](https://user-images.githubusercontent.com/1331477/210267838-6fd4f833-d0d8-4f2c-8206-8a3281c55a59.png)

This is very inconvenient:

- Some items are show on the top others are hidden in the categories, and there are even duplicates
- Initially one can't just scroll through if anything interesting is there
- Instead one needs to expand all items manually or know what to search for
- Even then, the primary order is the category and not the name of the project type one is filtering for

Instead if the "new project" style wizard is opened, it should use a flat list, sort it by name and probably only use the category to be shown next by the name in a shaded way.

**Comment by mickaelistria** [comment]
I think the category is extremely important to show to users as it allows them to discover what are the kinds of artifacts to deal with for a given framework. However, I like the proposal of flattening the list with the last part about showing category in a shaded way, and also in a way that filtering can also filter the category.

**Comment by ruspl-afed** [comment]
Could it be a radio control to switch content [and label] provider from the current foldered to a flat style?

**Comment by laeubi** [comment]
> Could it be a radio control to switch content [and label] provider from the current foldered to a flat style?

Everything is possible (even though a checkbox might be more suitable Ui control) but then somone needs to be dedicated to implement / support this as it requires additional effort e.g. to dynamically update the UI.

> I think the category is extremely important to show to users as it allows them to discover what are the kinds of artifacts to deal with for a given framework.

At least I have never be care much, this is more a technical than a user topic, e.g. "Java Project" is really clear enough what and for what this is performed, but as said I won't mind to show the category somewhere. I think there is even enough room to make it a TreeTable with one column is the name while the other is the category.

**Comment by mickaelistria** [comment]
> e.g. "Java Project" is really clear enough what and for what this is performed

Yes but it's one of the happy cases. Most content in "Plugin Development" is really hard to understand for a user without a category describing in which context those projects should be used; and the other way round, someone who just installed PDE would be happy to see a category allowing to identify the new available wizards.

Also, anticipating the next steps, if we want to apply something similar to the "New" wizard, then there are labels that make sense only in the context of the category, such as "C/C++ > Source file", "C/C++ > Header file" and also some labels that are identical and only disambiguised by the category; "Java > Class" vs "C/C++ > Class".
And I didn't check, but I imagine the Import/Export wizard have similar cases.
So the category is to be shown first to properly handle those cases.

> I think there is even enough room to make it a TreeTable with one column is the name while the other is the category.

I would personally prefer "[Java] Class" with "[Java]" being the usual decoration color (brownish on my machine) over columns, but I think one or the other would still be a good improvement for this wizard.

**Comment by laeubi** [comment]
Just to not been confused, there are two types of wizard dialogs:

1. "Projects..." (that only show wizards that create projects) and what I'm talking about here
2. "Others..." (that shows everything including create new files, and such stuff)

Here is a screenshot with expanded "Plugin-Devloment" how it is presented for the first case, if you click an item you see a description what this wizard does, if name+description is unclear to the usual user, I think this should be improved (independent of this proposal):
![grafik](https://user-images.githubusercontent.com/1331477/210321418-cf973831-c875-4abe-9d92-6ac24832c25f.png)

**Comment by iloveeclipse** [comment]
> Instead if the "new project" style wizard is opened, it should use a flat list, sort it by name and probably only use the category to be shown next by the name in a shaded way.

This doesn't scale and would be a step backward. In my IDE with CDT/Xtext and some other plugins installed, if I expand all the "new" categories I have to scroll few pages down to see entire list. Most of the entries in that flat list would be lost/meaningless out of context and would not be interesting for me (Xtext, EMF, C++ etc), but hiding the few in the category I'm interested in.

**Comment by mickaelistria** [comment]
> Just to not been confused, there are two types of wizard dialogs:

Actually, they all are instances of WorkbenchWizardSelectionPage which does provide the UI. And having different approaches to those different wizards would seem pretty inconsistent.

> if name+description is unclear to the usual user, I think this should be improved (independent of this proposal):

currently category+name+description is clear enough, and it's what all plugins have built on top of. You cannot suddenly say "we hide the category, now it's just name+description" because then it means this would be perceived as a regression to many users and a major annoyance to plugins who would need to react.
So category+name+description is still the thing to show, a way or another, and that mustn't be changed.

> This doesn't scale and would be a step backward

don't you use the filter?

**Comment by merks** [comment]
It's hard to imagine this list being flattened being usable:

![image](https://user-images.githubusercontent.com/208716/210327635-bb1838bf-4787-4db6-bfb5-5be3711af346.png)

Lots of examples:

![image](https://user-images.githubusercontent.com/208716/210327755-1a13c96c-c7e1-4af1-a451-da203b2705b8.png)

People tend to get very picky when you change something familiar to something "better" that is perhaps arguably not better but rather just different.

**Comment by laeubi** [comment]
> Actually, they all are instances of WorkbenchWizardSelectionPage which does provide the UI. And having different approaches to those different wizards would seem pretty inconsistent.

So we better show over-complicated complex hard to use wizard in all cases just for the sake of being consistently complex?

> This doesn't scale and would be a step backward.

Scale for what? For sure you can always install more and you will get more content, but having more categories do not help much.

>  if I expand all the "new" categories I have to scroll few pages down to see entire list.

The idea is that *usually* the list is not that long and the categories currently take up much space, e.g "java Delopment" only contains two items + one category, so it will only increase by one but still showing one more relevant (choosable) item.

Also the filtering should help to narrow things down to interesting items.

> It's hard to imagine this list being flattened being usable:

Its hard to believe that the current categorized tree makes it "usable" anyways, so what do we want to improve for:

1. Usual use cases
2. Use case where you blindly install anything you ever find on the internet to blow up your IDE

> Lots of examples:

Examples are already handled special by the wizard

**Comment by mickaelistria** [comment]
> So we better show over-complicated complex hard to use wizard in all cases just for the sake of being consistently complex?

The over-complication is proposing distinct interactions for very similar concepts. Then user has to learn more ways to interact. It takes a cognitive effort that is spared if we can keep things consistent.

**Comment by laeubi** [comment]
"Creating a new project" is similar to "creating something new" (even including in the later) but it does not mean that the same UI concepts suffice here. Beside that, not showing categories is nothing that should really overwhelm a user or requires "cognitive effort" to understand.

**Comment by iloveeclipse** [comment]
> The idea is that usually the list is not that long

Define "usually" in an extensible IDE used by millions of people in various ways. In my "usual" setup the list is over 3 pages long and will be not usable without categories.

**Comment by laeubi** [comment]
> > The idea is that usually the list is not that long
> 
> Define "usually" in an extensible IDE used by millions of people in various ways. In my "usual" setup the list is over 3 pages long and will be not usable without categories.

Usually for me is what can be found here: https://www.eclipse.org/downloads/packages/release/2022-12/r

Of course one can install more and more stuff added and then probably using the search bar will be more appropriate than scrolling the list, uninstall stuff not required anymore, or ask the project to allow more fine-grained control over features to add to the IDE.

Even though it is currently not possible, one might even want to hide things that are not needed/wanted somehow, but that would be a different proposal, maybe one can even think about having some kind of LRU sorting presenting things more often used first.

**Comment by ruspl-afed** [comment]
>  In my "usual" setup the list is over 3 pages long and will be not usable without categories.

To be honest, categories are just postponing a bit the problem of "long list" without solving it. 

Perhaps a tag-based filtering can help to find the right wizard. What about a line of contributable toggle buttons associated with tags?

**Comment by laeubi** [comment]
> > In my "usual" setup the list is over 3 pages long and will be not usable without categories.
> 
> To be honest, categories are just postponing a bit the problem of "long list" without solving it.
> 
> Perhaps a tag-based filtering can help to find the right wizard. What about a line of contributable toggle buttons associated with tags?

I my own RCP application I use [nebula chips](https://wiki.eclipse.org/Nebula_Chips) for fast category filtering. but I'm not sure if it would be ok to have nebula as a dependency for platform.

**Comment by mickaelistria** [comment]
>  I'm not sure if it would be ok to have nebula as a dependency for platform.

So far, Nebula is not a dependency, but I don't think there is a strict rule against it. But the expectations and limitations need to be discussed.
The issue with those drawn widgets (same with SWT C* widgets) are numerous: how do they react the theme change (ie if OS/e4 theme is changed, are nasty colors magically avoided), how do they work in term of accessibility (colorblind, screen readers, touchscreens, keyboard-only navigation), how do they like HiDPI, how do they perform, how can they be rendered in exotic cases.... Usually, when it comes to all those matters that are of high important for Platform, native widgets such as checkboxes are hard to beat and are always the safest choice.
